stdlib/LinearAlgebra/src/LinearAlgebra.jl

```diff
@@ -221,8 +221,8 @@ In this case, one must implement the strided array interface, which requires
 overrides of `strides(A::MyArray)` and `unknown_convert(::Type{Ptr{T}}, A::MyArray)`.
 """
 MemoryLayout(A::AbstractArray{T}) where {T} = UnknownLayout{T}()
-MemoryLayout(A::Vector{T}) where {T} = DenseLayout{T}()
-MemoryLayout(A::Matrix{T}) where {T} = DenseLayout{T}()
+MemoryLayout(A::DenseVector{T}) where {T} = DenseLayout{T}()
+MemoryLayout(A::DenseMatrix{T}) where {T} = DenseLayout{T}()
```

Comment:
This is wrong.  As I understand it, your `DenseLayout` means `DenseColumnMajor`.  This is not what `DenseMatrix` means — `DenseArray` just means you have `strides` and pointer conversion.  From the manual:

>DenseArray is an abstract subtype of AbstractArray intended to include all arrays that are laid out at regular offsets in memory, and which can therefore be passed to external C and Fortran functions expecting this memory layout. Subtypes should provide a method stride(A,k) that returns the "stride" of dimension k: increasing the index of dimension k by 1 should increase the index i of getindex(A,i) by stride(A,k). If a pointer conversion method Base.unsafe_convert(Ptr{T}, A) is provided, the memory layout should correspond in the same way to these strides.

Reply:
Thanks for letting me know. (Once tests are passing with the current names I'll clarify things by renaming `DenseLayout` -> `DenseColumnMajor` and `TransposeLayout` -> `DenseRowMajor`, or similar.)

Would you suggest removing the override for `DenseArray`s and instead implement it for each subtype (e.g. `SharedArray`) to make it work? 
```julia
LinearAlgebra.MemoryLayout(A::SharedArray{T}) where T = LinearAlgebra.DenseLayout{T}()
```

Are there any other `DenseArray`s that I should know about? I think `BitArray` should probably have its own layout... 

Also, for a row major matrix, would `stride(A,1)` be the row stride, or column stride?

Reply:
I'm a bit confused why column major/row major are different things: why is it not true that a matrix "column major" iff `stride(A,1) < stride(A,2)` and "row major" iff `stride(A,1) > stride(A,2)`?

Reply:
OK, I think I understand: based on [BLAS](https://software.intel.com/en-us/mkl-developer-reference-c-matrix-storage-schemes-for-blas-routines) `ColumnMajor` implies `stride(A,1) == 1` while `RowMajor` implies `stride(A,2) == 1`. Then `Strided` allows arbitrary strides.

BLAS/LAPACK only work with `ColumnMajor` or `RowMajor`. So I think these are the types that are needed:
```julia
abstract type DenseLayout{T} <: MemoryLayout{T} end
abstract type BlasMatrixLayout{T} <: DenseLayout{T} end 
struct RowMajorLayout{T} <: BlasMatrixLayout{T} end
struct ColumnMajorLayout{T} <: BlasMatrixLayout{T} end
struct ConjRowMajorLayout{T<:Complex} <: BlasMatrixLayout{T} end
struct ConjColumnMajorLayout{T<:Complex} <: BlasMatrixLayout{T} end
struct StridedLayout{T} <: DenseLayout{T} end
```